File: algorithms/search/binarySearch.ts
/**
 * Return the index of an element if exists, -1 otherwise
 */
function binarySearch(arr: number[], target: number): number {
    let left = 0, right = arr.length - 1
    while (left <= right) {
        const mid = left + Math.trunc((right - left) / 2)

        if (target === arr[mid]) {
            return mid
        }

        if (target < arr[mid]) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    return -1
}

export default binarySearch

File: algorithms/search/binary_search.go
package search

func BinarySearchInt(arr []int, target int) int {
	left, right := 0, len(arr)-1
	for left <= right {
		mid := left + (right-left)/2
		cur := arr[mid]

		if cur == target {
			return mid
		}

		if target < cur {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return -1
}
